### PR TITLE
RFC: Add a gem definition to the styleguides

### PR DESCRIPTION
**Pull request by boffbowsh**
This is intended as an easy way to add style guide compliance to GDS
Ruby projects, simply add it to the Gemfile and call `bundle exec
gds-ruby-styleguide`

Also bundled the Diff functionality from Whitehall
(https://github.com/alphagov/whitehall/blob/7f0ef8df4daa61a6e2ba6641c162
688717c7189f/lib/rubocop_diff.rb), accessible using the `--diff` and
`--cached` command line options

### CODE REVIEW

**Comment by dsingleton** [comment]
Having a gem to enforce the style guide easily and programatically seems like a good idea to me :+1: 

But i'm not sure if the styleguides repo becoming that gem is the right approach. :-1: 

The documentation here covers many languages/formats and it would be weird to have this repo be specifically a gem, and specifically for ruby-language checking (vs wrapping a SCSS ruby linter - hypothetically).

**Comment by boffbowsh** [comment]
I meant to say - it's entirely possible (I think?) to create a wrapper in this repo too for flake8 / jshint / csslint. Notice I've versioned the repo as a whole also, means you can say "Oh project X is only compliant with 1.2.x of the styleguide, not 1.3.x, we should fix that. Most apps should have lintable things in Ruby, JS and CSS, so imo one repo is a good way to go.

**Comment by boffbowsh** [comment]
To clarify, I mean we should version both the English documentation and the compliancy configs that relate to it.

**Comment by dsingleton** [comment]
Yep, it makes sense to keep the human doc and code doc in sync. 

I meant more the packaging gubbins, all the package.json, gemspec, etc etc. This can get pretty messy/noisy once you add a few packaging systems. The approach for govuk_template is to have a core repo which contains the authoritative bits and a series of automatically built/published when the core repo changes, see: https://github.com/search?utf8=%E2%9C%93&q=govuk_template_+%40alphagov&type=Repositories&ref=searchresults

That would probably be overkill for what you're trying to do here. So maybe we'd be better off just just having it all in here until it becomes unwieldy and extracting it out then.

**Comment by boffbowsh** [comment]
Ah I see. Yeah I think this is probably good as a first go. See if it gains traction first and actually solves a problem.

**Comment by dsingleton** [comment]
:+1: in which case, maybe making the naming a bit vaguer, so it could cover CSS/JS linting as part of this gem would give flexibility before having to split things out?

**Comment by tijmenb** [comment]
This seems like a good idea to me. :+1: 

Could we make it work with `rubocop --auto-gen-config`? That generates a `.rubocop_todo.yml` with all violations, so you can solve them one by one or make explicit exceptions for the project.

**Comment by boffbowsh** [comment]
CSS/JS would need to be covered by some sort of npm package, the linters can't be installed by a Gem. But yeah, I guess it makes sense to know that it's always just the `gds-styleguide` package you want, be that gem/npm/pip etc.

@tijmenb probably, but I'll be honest I'm not a big fan of that workflow. If your codebase isn't ready for the full Rubocop treatment, use the `--diff` option in Jenkins. I'm happy for others to weigh in on that issue though.

**Comment by annashipman** [comment]
I also think it's a bit odd for this to be a gem, especially as it includes guidance for things like commit messages, pull requests etc, not just other languages. But if @bradleywright thinks this is a good idea then I'm not going to argue the point.

Few specific comments if this does become the way to go, I will comment on the diff.

File: gds-ruby-styleguide.gemspec [hunk 1]
```diff
@@ -0,0 +1,22 @@
+# coding: utf-8
+lib = File.expand_path('../lib', __FILE__)
+$LOAD_PATH.unshift(lib) unless $LOAD_PATH.include?(lib)
+
+Gem::Specification.new do |spec|
+  spec.name          = "gds-ruby-styleguide"
```
Comment (annashipman) [inline gds-ruby-styleguide.gemspec]: It's the GOV.UK styleguide, not the GDS styleguide.

File: gds-ruby-styleguide.gemspec [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+  spec.version       = File.read(File.expand_path('../VERSION', __FILE__)).strip
+  spec.authors       = ["Paul Bowsher"]
```
Comment (annashipman) [inline gds-ruby-styleguide.gemspec]: This should be Government Digital Service.

File: gds-ruby-styleguide.gemspec [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+  spec.email         = ["paul.bowsher@gmail.com"]
```
Comment (annashipman) [inline gds-ruby-styleguide.gemspec]: As above.

File: gds-ruby-styleguide.gemspec [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+  spec.summary       = %q{A wrapper around rubocop, configured with the GDS style guides}
+  spec.description   = %q{Include this in your project to easily validate it against the GDS style guides}
```
Comment (annashipman) [inline gds-ruby-styleguide.gemspec]: Two lines above, GOV.UK not GDS.

**Comment by edds** [comment]
I am really not a fan of this repo becoming the gem. I think it is a great idea to have the style guide automatically enforced, but that should be separate from the thinking and reasoning. I want to watch this repo to see changes in how I should write my code, not to see version bumps of ruby gems and gem fixes. Also having a `VERSION` file in the root of the project is super disingenuous as that implies we are versioning our style guide, which we aren't.

**Comment by boffbowsh** [comment]
Ok, consensus seems to be to move this out into its own (correctly named) repo. But overall, is having this wrapper gem worth it or no?

**Comment by bradleywright** [comment]
I think we should be linting all our code as much as we can, and getting automatic feedback about it in our test suites. I don't care much about the implementation, just as long as we're linting the code and reporting on it.

**Comment by boffbowsh** [comment]
So then, shall we have a `govuk-lint` repo which contains configs for rubocop, flake8, jshint, scss-lint and has scripts like `govuk-lint-ruby`, `govuk-lint-js` etc which wrap the linters and use the correct configs?

**Comment by dsingleton** [comment]
> So then, shall we have a govuk-lint repo which contains configs for rubocop, flake8, jshint, scss-lint and has scripts like govuk-lint-ruby, govuk-lint-js etc which wrap the linters and use the correct configs?

That sounds reasonable to me. Just thinking through workflow, changes to the styleguide that require changing a config are going to require two PRs, which may be a bit painful, but i'm not sure there's a less painful option. Lets see how it goes.

**Pull request closed without merging**